File: scripts/extract_pdf_images.py
#!/usr/bin/env python3
"""
Extract PDF pages as images and upload to Missira's artworks
Requires: pip install pdf2image pillow requests
"""

import os
import sys
import json
import base64
import requests
from pathlib import Path

# Try imports
try:
    from pdf2image import convert_from_path
    from PIL import Image
    import io
except ImportError as e:
    print(f"❌ Missing dependency: {e}")
    print("\n📦 Install with:")
    print("   pip install pdf2image pillow requests")
    sys.exit(1)

# Config
PDF_PATH = r"C:\Users\Moctar Sidibe\Desktop\_Catalogue MISS'ART_compressed.pdf"
SUPABASE_URL = "https://wyrmpddlhldjzoiwbshj.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Ind5cm1wZGRsaGxkanpvaXdic2hqIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3MjcwNjA0NywiZXhwIjoyMDg4MjgyMDQ3fQ.QQ6Bt-Tk5Vm_Lp8zD1SewXFgPRHGZgvA2v6zj87ijqs"
ARTIST_ID = "f1f5ea0f-fce3-45d8-a770-4c3297cee4e1"

def extract_and_upload():
    print("📄 EXTRACTING PDF IMAGES\n")

    # Check PDF exists
    if not os.path.exists(PDF_PATH):
        print(f"❌ PDF not found: {PDF_PATH}")
        return

    print(f"✅ PDF found: {PDF_PATH}\n")

    # Convert PDF pages to images
    print("🔄 Converting PDF pages to images...\n")
    try:
        images = convert_from_path(PDF_PATH, first_page=1, last_page=9)
        print(f"✅ Extracted {len(images)} pages\n")
    except Exception as e:
        print(f"❌ Extraction error: {e}")
        return

    # Get artworks from Supabase
    print("🔍 Fetching artworks...\n")
    headers = {
        "apikey": SUPABASE_KEY,
        "Authorization": f"Bearer {SUPABASE_KEY}",
    }

    response = requests.get(
        f"{SUPABASE_URL}/rest/v1/artworks?"
        f"artist_id=eq.{ARTIST_ID}&select=id,title&order=created_at.desc",
        headers=headers,
    )

    if response.status_code != 200:
        print(f"❌ Failed to fetch artworks: {response.text}")
        return

    artworks = response.json()
    print(f"✅ Found {len(artworks)} artworks\n")

    # Upload images
    print("📤 Uploading images...\n")
    uploaded = 0

    for i, artwork in enumerate(artworks[: len(images)]):
        print(f"   [{i+1}/{len(artworks)}] {artwork['title']}...", end="", flush=True)

        try:
            # Convert image to JPEG
            img = images[i]
            img_resized = img.resize((800, 600), Image.Resampling.LANCZOS)

            # Convert to base64
            img_bytes = io.BytesIO()
            img_resized.save(img_bytes, format="JPEG", quality=85)
            img_base64 = base64.b64encode(img_bytes.getvalue()).decode()
            data_uri = f"data:image/jpeg;base64,{img_base64}"

            # Update artwork
            update_response = requests.patch(
                f"{SUPABASE_URL}/rest/v1/artworks?id=eq.{artwork['id']}",
                json={"image": data_uri},
                headers=headers,
            )

            if update_response.status_code == 204:
                print(" ✅")
                uploaded += 1
            else:
                print(f" ❌ ({update_response.status_code})")
        except Exception as e:
            print(f" ❌ ({e})")

    print(f"\n✅ Uploaded: {uploaded}/{len(artworks)}")
    print("\n🔄 Next: Hard refresh (Ctrl+F5) to see PDF images!")

if __name__ == "__main__":
    extract_and_upload()

File: scripts/extract-pdf-images.mjs
#!/usr/bin/env node

/**
 * Extract images from Missira's PDF catalogue using pdf.js
 */

import * as fs from "fs";
import * as path from "path";
import * as pdfjsLib from "pdfjs-dist";

const { createClient } = await import("@supabase/supabase-js");
const { Canvas, createCanvas } = await import("canvas").catch(() => null);

const SUPABASE_URL = "https://wyrmpddlhldjzoiwbshj.supabase.co";
const SUPABASE_SERVICE_ROLE_KEY =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Ind5cm1wZGRsaGxkanpvaXdic2hqIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3MjcwNjA0NywiZXhwIjoyMDg4MjgyMDQ3fQ.QQ6Bt-Tk5Vm_Lp8zD1SewXFgPRHGZgvA2v6zj87ijqs";

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY);
const ARTIST_ID = "f1f5ea0f-fce3-45d8-a770-4c3297cee4e1";
const PDF_PATH = path.join(path.dirname("."), "..", "Desktop", "_Catalogue MISS'ART_compressed.pdf");

// Set worker
pdfjsLib.GlobalWorkerOptions.workerSrc = `//cdn.jsdelivr.net/npm/pdfjs-dist@${pdfjsLib.version}/build/pdf.worker.min.js`;

async function extractAndUpload() {
  console.log("📄 EXTRACTING IMAGES FROM PDF\n");

  try {
    if (!fs.existsSync(PDF_PATH)) {
      console.log(`❌ PDF not found: ${PDF_PATH}`);
      return;
    }

    console.log(`✅ PDF found: ${PDF_PATH}\n`);

    // Load PDF
    const pdfBuffer = fs.readFileSync(PDF_PATH);
    const pdfData = await pdfjsLib.getDocument({ data: pdfBuffer }).promise;
    const numPages = pdfData.numPages;

    console.log(`📊 PDF has ${numPages} pages\n`);

    // Get artworks
    const { data: artworks } = await supabase
      .from("artworks")
      .select("id, title")
      .eq("artist_id", ARTIST_ID)
      .order("created_at", { ascending: false });

    if (!artworks) {
      console.log("❌ No artworks found");
      return;
    }

    console.log(`Extracting ${Math.min(artworks.length, numPages)} page(s)...\n`);

    // Extract pages as images
    let uploaded = 0;
    for (let pageNum = 1; pageNum <= Math.min(artworks.length, numPages); pageNum++) {
      const artwork = artworks[pageNum - 1];

      process.stdout.write(`   [${pageNum}/${artworks.length}] ${artwork.title}...`);

      try {
        const page = await pdfData.getPage(pageNum);
        const viewport = page.getViewport({ scale: 2 });

        // Render page to canvas (if canvas available)
        if (createCanvas) {
          const canvas = createCanvas(viewport.width, viewport.height);
          const context = canvas.getContext("2d");

          const renderContext = {
            canvasContext: context,
            viewport: viewport,
          };

          await page.render(renderContext).promise;
          const imageBuffer = canvas.toBuffer("image/jpeg");
          const dataUri = `data:image/jpeg;base64,${imageBuffer.toString("base64")}`;

          // Update artwork
          const { error } = await supabase
            .from("artworks")
            .update({ image: dataUri })
            .eq("id", artwork.id);

          if (error) {
            console.log(` ❌`);
          } else {
            console.log(` ✅`);
            uploaded++;
          }
        } else {
          // Fallback: skip if canvas not available
          console.log(` ⏭️  (canvas not available)`);
        }
      } catch (err) {
        console.log(` ❌ ${err.message}`);
      }
    }

    console.log(`\n✅ Uploaded: ${uploaded}/${artworks.length}`);
    console.log("\n🔄 Next: Hard refresh (Ctrl+F5) to see PDF images!");
  } catch (err) {
    console.error("❌ ERROR:", err.message);
  }
}

extractAndUpload();
